File: sdk/utils/assetsTracker.ts
import { ObjectId, Collection, Db } from "mongodb";
import { logger } from "../utils"; 

/**
 * SDK for tracking and cleaning up MongoDB assets.
 */
export class AssetsTracker {
    private trackedAssets: Map<string, ObjectId[]> = new Map();
    private dbInstance: Db; 

    constructor(db: Db) {
        this.dbInstance = db;
    }

    public track(params: Record<string, ObjectId>): void {
        const collectionName = Object.keys(params)[0];
        const assetId = params[collectionName];

        if (!collectionName || !assetId) {
            logger.error("[DbAssetsSDK.track]: Invalid input");
            return;
        }

        if (!this.trackedAssets.has(collectionName)) {
            this.trackedAssets.set(collectionName, []);
        }

        this.trackedAssets.get(collectionName)!.push(assetId);
        logger.debug(`[DbAssetsSDK.track]: Tracked ${collectionName} ID ${assetId.toHexString()}`);
    }

    public async cleanup(collectionsToCleanup: Record<string, boolean>): Promise<void> {
        const collections = Object.keys(collectionsToCleanup).filter(c => collectionsToCleanup[c]);
        if (!collections.length) {
            logger.warn("[DbAssetsSDK.cleanup]: No collections specified for cleanup.");
            return;
        }

        for (const collectionName of collections) {
            const ids = this.trackedAssets.get(collectionName);
            if (!ids || ids.length === 0) {
                logger.warn(`[DbAssetsSDK.cleanup]: No tracked assets in ${collectionName}`);
                continue;
            }

            try {
                const collection: Collection = this.dbInstance.collection(collectionName);
                const result = await collection.deleteMany({ _id: { $in: ids } });
                logger.info(`[DbAssetsSDK.cleanup]: Deleted ${result.deletedCount}/${ids.length} from ${collectionName}`);
            } catch (err) {
                logger.error(`[DbAssetsSDK.cleanup]: Error cleaning ${collectionName}: ${err}`);
            }

            this.trackedAssets.set(collectionName, []);
        }

        logger.info("[DbAssetsSDK.cleanup]: Cleanup complete.");
    }
}

File: playwright_py/utils/AssetsTracker.py
from typing import Dict, List
from bson.objectid import ObjectId
from pymongo.database import Database
from models import collections, CollectionsType
import asyncio
import logging
logger = logging.getLogger(__name__)


class AssetsTracker:
    """
    Tracks and cleans up MongoDB assets created during tests.
    Only accepts keys defined in CollectionsType.
    """

    def __init__(self, db: Database):
        self._db_instance: Database = db
        self._tracked_assets: Dict[CollectionsType, List[ObjectId]] = {}

    async def track(self, params: Dict[CollectionsType, ObjectId]) -> None:
        """
        Track an asset by collection name (must be in CollectionsType) and its ObjectId.
        """
        for collection_name, asset_id in params.items():
            if collection_name not in self._tracked_assets:
                self._tracked_assets[collection_name] = []
            self._tracked_assets[collection_name].append(asset_id)
            logger.debug("Tracked asset ID %s in collection %s", asset_id, collection_name)

    async def cleanup(self, collections_to_cleanup: Dict[CollectionsType, bool]) -> None:
        """
        Cleanup tracked assets for the specified collections.
        """
        collections = [c for c, should_cleanup in collections_to_cleanup.items() if should_cleanup]

        if not collections:
            logger.warning("No collections specified for cleanup.")
            return

        for collection_name in collections:
            ids = self._tracked_assets.get(collection_name, [])
            if not ids:
                logger.warning("No tracked assets found in collection: %s", collection_name)
                continue

            try:
                collection = self._db_instance.get_collection(collection_name)
                result = await asyncio.to_thread(
                    collection.delete_many,
                    {"_id": {"$in": ids}}
                )
                logger.info(
                    "Deleted %d/%d assets from collection %s",
                    result.deleted_count,
                    len(ids),
                    collection_name
                )
            except Exception as e:
                logger.error("Error cleaning collection %s: %s", collection_name, e, exc_info=True)

            self._tracked_assets[collection_name] = []

        logger.info("Cleanup complete.")
